packages/react-utilities/src/hooks/useControllableState.ts

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import { useConst } from './useConst';
+
+type NonUndefined<T> = T extends undefined ? never : T;
+
+type UseAutoControlledOptions<ControllableState, State> = {
+  /**
+   * User provided default state or factory initializer
+   */
+  defaultState?: State | (() => State);
+  /**
+   * User provided controllable state, undefined state means internal state will be used
+   */
+  state: ControllableState;
+  /**
+   * Used to initialize state if all user provided states are undefined
+   */
+  initialState: State;
+};
+
+function isUndefined(state: unknown): state is undefined {
+  return typeof state === 'undefined';
+}
+
+function isFactoryDispatch<State>(newState: React.SetStateAction<State>): newState is (prevState: State) => State {
+  return typeof newState === 'function';
+}
+
+/**
+ * A useState 'like' hook that allows optional user control
+ * Useful for components which allow uncontrolled and controlled behaviours for users
+ * @returns - https://reactjs.org/docs/hooks-state.html
+ */
+export const useControllableState = <ControllableState, State extends NonUndefined<ControllableState>>(
+  options: UseAutoControlledOptions<ControllableState, State>,
+): [State, React.Dispatch<React.SetStateAction<State>>] => {
+  const isControlled = useIsControlled(options.state);
+  const initialState = isUndefined(options.defaultState) ? options.initialState : options.defaultState;
+  const [internalState, setInternalState] = React.useState<State>(initialState);
+
+  const state = isControlled ? (options.state as NonUndefined<State>) : internalState;
+
+  const stateRef = React.useRef(state);
+  React.useEffect(() => {
+    stateRef.current = state;
+  }, [state]);
+
+  const setState = React.useCallback((newState: React.SetStateAction<State>) => {
```

Comment:
```suggestion
  // To match the behavior of the setter returned by React.useState, this callback's identity
  // should never change. This means it MUST NOT directly reference variables that can change.
  const setState = React.useCallback((newState: React.SetStateAction<State>) => {
```

I think that a comment from original `useControllableValue` is useful